<a href="https://colab.research.google.com/github/NataliaVrabcova/MS02700_workshop/blob/master/healtcare_assesment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!apt-get install git -y

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
git is already the newest version (1:2.34.1-1ubuntu1.11).
0 upgraded, 0 newly installed, 0 to remove and 49 not upgraded.


In [ ]:

!git clone https://your-username:AwAbEdrG7ZMbGxoSYAY3@labcode.mdx.ac.uk/nv204/assessment-1-neural-networks.git

Cloning into 'assessment-1-neural-networks'...
remote: Enumerating objects: 22, done.
remote: Counting objects: 100% (22/22), done.
remote: Compressing objects: 100% (19/19), done.
remote: Total 22 (delta 2), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (22/22), 2.01 MiB | 2.05 MiB/s, done.
Resolving deltas: 100% (2/2), done.


In [ ]:
!rm -rf assessment-1-neural-networks

In [ ]:
%cd /content/assessment-1-neural-networks/notebooks


/content/assessment-1-neural-networks/notebooks


In [ ]:
!git config --global user.email "nv204@live.mdx.ac.uk"
!git config --global user.name "NV204"

In [ ]:
!git add healtcare_assesment1.ipynb



In [ ]:
import pandas as pd

# Adjust the path to match your repository structure
data = pd.read_csv('/content/assessment-1-neural-networks/data/healthcare_noshows_appointments.csv')

# Verify the dataset is loaded
print(data.head())

      PatientId  AppointmentID Gender ScheduledDay AppointmentDay  Age  \
0  2.987250e+13        5642903      F   2016-04-29     2016-04-29   62   
1  5.589978e+14        5642503      M   2016-04-29     2016-04-29   56   
2  4.262962e+12        5642549      F   2016-04-29     2016-04-29   62   
3  8.679512e+11        5642828      F   2016-04-29     2016-04-29    8   
4  8.841186e+12        5642494      F   2016-04-29     2016-04-29   56   

       Neighbourhood  Scholarship  Hipertension  Diabetes  Alcoholism  \
0    JARDIM DA PENHA        False          True     False       False   
1    JARDIM DA PENHA        False         False     False       False   
2      MATA DA PRAIA        False         False     False       False   
3  PONTAL DE CAMBURI        False         False     False       False   
4    JARDIM DA PENHA        False          True      True       False   

   Handcap  SMS_received  Showed_up  Date.diff  
0    False         False       True          0  
1    False        

In [ ]:
# Importing libraries
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
from keras.models import Sequential
from keras.layers import Dense, Dropout, Input
from keras.optimizers import Adam


In [ ]:
!git commit -m "Updated libraries"





[master (root-commit) bc38a05] Updated libraries
 1 file changed, 1 insertion(+)
 create mode 100644 healtcare_assesment1.ipynb


Preprocessing the Data

In [ ]:
# Selecting features and target variable
X = data.drop(['Showed_up', 'PatientId', 'AppointmentID'], axis=1)
y = data['Showed_up']

# Encoding categorical variables
X = pd.get_dummies(X, drop_first=True)

# Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Feature scaling
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

Defining the architecture of the neural network:

In [ ]:
# Defining the neural network model
model = Sequential([
    Input(shape=(X_train.shape[1],)),
    Dense(64, activation='relu'),
    Dropout(0.5),
    Dense(32, activation='relu'),
    Dropout(0.5),
    Dense(1, activation='sigmoid')
])

# Compile the model
model.compile(loss='binary_crossentropy', optimizer=Adam(), metrics=['accuracy'])


In [ ]:

!git commit -m "Defined the neural network architecture"


On branch master
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.gitkeep

nothing added to commit but untracked files present (use "git add" to track)


Train the model with your preprocessed data:

In [ ]:
# Train the model
history = model.fit(X_train, y_train, epochs=50, batch_size=32, validation_data=(X_test, y_test))


Epoch 1/50
2675/2675 ━━━━━━━━━━━━━━━━━━━━ 8s 2ms/step - accuracy: 0.7811 - loss: 0.5376 - val_accuracy: 0.7978 - val_loss: 0.4707
Epoch 2/50
2675/2675 ━━━━━━━━━━━━━━━━━━━━ 9s 2ms/step - accuracy: 0.7954 - loss: 0.4736 - val_accuracy: 0.7979 - val_loss: 0.4591
Epoch 3/50
2675/2675 ━━━━━━━━━━━━━━━━━━━━ 6s 2ms/step - accuracy: 0.7962 - loss: 0.4640 - val_accuracy: 0.7979 - val_loss: 0.4565
Epoch 4/50
2675/2675 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.7977 - loss: 0.4573 - val_accuracy: 0.7981 - val_loss: 0.4528
Epoch 5/50
2675/2675 ━━━━━━━━━━━━━━━━━━━━ 8s 2ms/step - accuracy: 0.7980 - loss: 0.4529 - val_accuracy: 0.7984 - val_loss: 0.4506
Epoch 6/50
2675/2675 ━━━━━━━━━━━━━━━━━━━━ 7s 3ms/step - accuracy: 0.7972 - loss: 0.4517 - val_accuracy: 0.7984 - val_loss: 0.4501
Epoch 7/50
2675/2675 ━━━━━━━━━━━━━━━━━━━━ 6s 2ms/step - accuracy: 0.7946 - loss: 0.4527 - val_accuracy: 0.7979 - val_loss: 0.4499
Epoch 8/50
2675/2675 ━━━━━━━━━━━━━━━━━━━━ 10s 2ms/step - accuracy: 0.7974 - loss: 0.4502 

Evaluate the model’s performance:

In [ ]:
# Evaluate the model
y_pred = (model.predict(X_test) > 0.5).astype(int)
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.4f}")

669/669 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step
Accuracy: 0.7991


In [ ]:

!git commit -m "Trained the model and then evaluated it's performance"



On branch master
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.gitkeep

nothing added to commit but untracked files present (use "git add" to track)


Evaluate the model on the test set

In [ ]:

test_loss, test_accuracy = model.evaluate(X_test, y_test)

print(f"Test Loss: {test_loss}")
print(f"Test Accuracy: {test_accuracy}")

669/669 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7976 - loss: 0.4506
Test Loss: 0.45027637481689453
Test Accuracy: 0.7991400957107544


Test Loss: 0.4506
  This value indicates the average error (or loss) of the model when it made predictions on the test data. A lower loss generally means that the model made fewer mistakes.

Test Accuracy: 0.7991 (or 79.91%)
  This metric shows the proportion of correct predictions (true positives + true negatives) out of the total predictions on the test set. A value closer to 1 indicates better performance.

The model has an accuracy of around 79.91% on the test data, which suggests it performs reasonably well at predicting whether a patient will show up for their appointment. This is a good starting point

In [ ]:
!git commit -m "Evaluating the Model"

On branch master
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.gitkeep

nothing added to commit but untracked files present (use "git add" to track)


3) Part 3: Mathematical Analysis

Define Weight Matrices and Bias Vectors:

 We initialize the weight matrices W1, W2, and W3 with random values.  We also define the corresponding biases b1, b2, and b3.

Forward Propagation:
  Input Layer: Compute the weighted sum for the input data.
 Hidden Layer 1: Apply the ReLU activation.
 Hidden Layer 2: Apply the ReLU activation.
 Output Layer: Apply the Sigmoid activation to get the output as a probability.

      PatientId  AppointmentID Gender ScheduledDay AppointmentDay  Age  \
0  2.987250e+13        5642903      F   2016-04-29     2016-04-29   62   
1  5.589978e+14        5642503      M   2016-04-29     2016-04-29   56   
2  4.262962e+12        5642549      F   2016-04-29     2016-04-29   62   
3  8.679512e+11        5642828      F   2016-04-29     2016-04-29    8   
4  8.841186e+12        5642494      F   2016-04-29     2016-04-29   56   

       Neighbourhood  Scholarship  Hipertension  Diabetes  Alcoholism  \
0    JARDIM DA PENHA        False          True     False       False   
1    JARDIM DA PENHA        False         False     False       False   
2      MATA DA PRAIA        False         False     False       False   
3  PONTAL DE CAMBURI        False         False     False       False   
4    JARDIM DA PENHA        False          True      True       False   

   Handcap  SMS_received  Showed_up  Date.diff  
0    False         False       True          0  
1    False        

In [ ]:
import numpy as np
# Load the dataset
df = data


# Convert categorical variables to numerical format using pd.get_dummies
X = pd.get_dummies(df.drop(columns=['Showed_up']))  # Features with categorical encoding
y = df['Showed_up'].values  # Target

# Convert to numpy arrays
X = X.values

# Define the input dimensions
m, n = X.shape

# Define the number of neurons in each layer
hidden_neurons_1 = 64
hidden_neurons_2 = 32
output_neurons = 1

# Randomly initialize weights and biases
np.random.seed(42)  # For reproducibility
W1 = np.random.rand(n, hidden_neurons_1)  # Input to Hidden Layer 1
b1 = np.random.rand(hidden_neurons_1)     # Bias for Hidden Layer 1

W2 = np.random.rand(hidden_neurons_1, hidden_neurons_2)  # Hidden Layer 1 to Hidden Layer 2
b2 = np.random.rand(hidden_neurons_2)                   # Bias for Hidden Layer 2

W3 = np.random.rand(hidden_neurons_2, output_neurons)  # Hidden Layer 2 to Output Layer
b3 = np.random.rand(output_neurons)                    # Bias for Output Layer

# Forward propagation through Hidden Layer 1
Z1 = np.dot(X, W1) + b1  # Weighted sum
A1 = np.maximum(0, Z1)   # ReLU activation

# Forward propagation through Hidden Layer 2
Z2 = np.dot(A1, W2) + b2  # Weighted sum
A2 = np.maximum(0, Z2)   # ReLU activation

# Forward propagation through Output Layer
Z3 = np.dot(A2, W3) + b3  # Weighted sum
A3 = 1 / (1 + np.exp(-Z3))  # Sigmoid activation

print("Input Layer Output:", X)
print("Hidden Layer 1 Output:", A1)
print("Hidden Layer 2 Output:", A2)
print("Output Layer Output:", A3)

KeyboardInterrupt: 